art/defences/trainer/adversarial_trainier_madry_pgd.py

```diff
@@ -0,0 +1,99 @@
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements adversarial training following Madry's Protocol.
+
+| Paper link: https://arxiv.org/abs/1706.06083
+
+| Please keep in mind the limitations of defences. While adversarial training is widely regarded as a promising,
+principled approach to making classifiers more robust (see https://arxiv.org/abs/1802.00420), very careful evaluations
+are required to assess its effectiveness case by case (see https://arxiv.org/abs/1902.06705).
+"""
+import logging
+
+import numpy as np
+
+from art.attacks.projected_gradient_descent import ProjectedGradientDescent
+from art.defences.adversarial_trainer import AdversarialTrainer
+
+logger = logging.getLogger(__name__)
+
+
+class AdversarialTrainerMadryPGD(AdversarialTrainer):
+    """
+    Class performing adversarial training following Madry's Protocol.
+
+    | Paper link: https://arxiv.org/abs/1706.06083
+
+    | Please keep in mind the limitations of defences. While adversarial training is widely regarded as a promising,
+    principled approach to making classifiers more robust (see https://arxiv.org/abs/1802.00420), very careful
+    evaluations are required to assess its effectiveness case by case (see https://arxiv.org/abs/1902.06705).
+    """
+
+    def __init__(self, classifier, max_num_training_steps=80000, batch_size=128, eps=8, eps_step=2, max_iter=7,
+                 random_init=True, **kwargs):
+        """
+        Create an :class:`.AdversarialTrainerMadryPGD` instance.
+
+        Default values are for CIFAR-10 in pixel range 0-255.
+
+        :param classifier: Classifier to train adversarially.
+        :type classifier: :class:`.Classifier`
+        :param batch_size: Size of the batch on which adversarial samples are generated.
+        :type batch_size: `int`
+        :param eps: Maximum perturbation that the attacker can introduce.
+        :type eps: `float`
+        :param eps_step: Attack step size (input variation) at each iteration.
+        :type eps_step: `float`
+        :param max_iter: The maximum number of iterations.
+        :type max_iter: `int`
+        :param num_random_init: Number of random initialisations within the epsilon ball. For num_random_init=0
+            starting at the original input.
+        :type num_random_init: `int`
+        """
+
+        self.max_num_training_steps = max_num_training_steps
+        self.batch_size = batch_size
+        self.nb_epochs = int(max_num_training_steps / (50000 / batch_size))
```

Comment:
Where does the `50000` come from? Supposedly this is the length of the training set. Perhaps it's easier for the user to directly provide the number of epochs, or is `max_num_training_steps` used anywhere else?

Reply:
I think the 50000 came with the source code. After our discussion I replaced it with the argument `nb_epochs` in 661c27dfd27cb4c079dbf0959e8b1700e35891b9.